File: gui/android/app/src/main/java/fr/cyclooo/opendeezer/engine/Engine.kt
package fr.cyclooo.opendeezer.engine

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import odmobile.Odmobile
import org.json.JSONObject

/**
 * Thin Kotlin facade over the gomobile-bound [Odmobile] static methods.
 *
 * Note: gomobile maps Go `int` to Java `long`, so engine getters such as
 * [state]/[quality]/[crossfadeMs]/[finishedCount] return Long and the
 * corresponding setters take Long arguments. Everything that touches the
 * network is exposed as a suspend function pinned to [Dispatchers.IO].
 */
object Engine {

    // Player states (mirror audio.State in the Go core).
    const val STOPPED = 0
    const val LOADING = 1
    const val PLAYING = 2
    const val PAUSED = 3
    const val ERROR = 4

    // ---- lifecycle / account ----

    suspend fun init(arl: String): Boolean = withContext(Dispatchers.IO) {
        try {
            Odmobile.init(arl)
        } catch (_: Throwable) {
            false
        }
    }

    fun loggedIn(): Boolean = runCatching { Odmobile.loggedIn() }.getOrDefault(false)

    // Why the most recent [init] failed: 0 ok, 1 ARL expired/invalid, 2 no
    // internet, 3 other. Only meaningful right after [init] returned false.
    // gomobile maps Go int -> Java long, hence toInt(); defaults to 3 (other).
    fun loginErrorKind(): Int = runCatching { Odmobile.loginErrorKind().toInt() }.getOrDefault(3)

    suspend fun account(): Account? = withContext(Dispatchers.IO) {
        Json.account(runCatching { Odmobile.account() }.getOrNull())
    }

    fun setClientInfo(client: String, device: String) {
        runCatching { Odmobile.setClientInfo(client, device) }
    }

    // Tears the engine session down: stops playback, closes the control server
    // and Connect-host advertising, and forgets the Deezer client so a later
    // init starts fresh. Suspend — it closes network listeners.
    suspend fun logout() = io { runCatching { Odmobile.logout() }.let {} }

    // Checks GitHub for a newer release. Network-bound and best-effort — a
    // failure (offline, rate-limited, etc.) just yields null, never a crash.
    suspend fun checkUpdate(): UpdateInfo? = io { Json.updateInfo(runCatching { Odmobile.checkUpdate() }.getOrNull()) }

    // ---- browse (all network; return parsed models) ----

    suspend fun favorites(): List<Track> = io { Json.tracks(Odmobile.favorites()) }

    // Client-side favourite check: the engine has no dedicated lookup, so we
    // reconcile against the loaded favourites list. Best-effort — false on error.
    suspend fun isFavorite(id: String): Boolean =
        runCatching { favorites().any { it.id == id } }.getOrDefault(false)

    // The account's liked track ids in a single call (reuses the same favourites
    // fetch the library view uses), so hearts render truthfully across lists
    // without a per-track lookup. Empty on error / not logged in.
    suspend fun favoriteIds(): Set<String> =
        io { Json.favoriteIds(runCatching { Odmobile.favoriteIDsJSON() }.getOrNull()) }
    suspend fun playlists(): List<Playlist> = io { Json.playlists(Odmobile.playlists()) }
    suspend fun playlistTracks(id: String): List<Track> = io { Json.tracks(Odmobile.playlistTracks(id)) }
    suspend fun albumTracks(id: String): List<Track> = io { Json.tracks(Odmobile.albumTracks(id)) }
    suspend fun flow(): List<Track> = io { Json.tracks(Odmobile.flow()) }
    suspend fun artistTop(id: String): List<Track> = io { Json.tracks(Odmobile.artistTop(id)) }

    // Full artist profile (top tracks + albums + related artists) in one call.
    // Null when the engine returned an error payload so the screen can offer a retry.
    suspend fun artistProfile(id: String): ArtistPage? =
        io { Json.artistPage(runCatching { Odmobile.artistProfile(id) }.getOrNull()) }
    suspend fun search(q: String): SearchResults = io { Json.search(Odmobile.search(q)) }
    suspend fun charts(): SearchResults = io { Json.search(Odmobile.charts()) }
    suspend fun home(): HomeData = io { Json.home(runCatching { Odmobile.home() }.getOrNull()) }
    suspend fun searchPodcasts(q: String): List<Podcast> = io { Json.podcasts(Odmobile.searchPodcasts(q)) }
    suspend fun podcastEpisodes(id: String): List<Episode> = io { Json.episodes(Odmobile.podcastEpisodes(id)) }
    suspend fun lyrics(id: String): Lyrics = io { Json.lyrics(Odmobile.lyrics(id)) }

    // Failure-aware variants: null when the engine returned an {"error":...}
    // payload (or nothing), so "couldn't load" is distinguishable from "empty"
    // and retry UIs can actually trigger.
    suspend fun homeOrNull(): HomeData? =
        io { raw(runCatching { Odmobile.home() }.getOrNull())?.let(Json::home) }
    suspend fun searchOrNull(q: String): SearchResults? =
        io { raw(runCatching { Odmobile.search(q) }.getOrNull())?.let(Json::search) }
    suspend fun chartsOrNull(): SearchResults? =
        io { raw(runCatching { Odmobile.charts() }.getOrNull())?.let(Json::search) }
    suspend fun flowOrNull(): List<Track>? =
        io { raw(runCatching { Odmobile.flow() }.getOrNull())?.let(Json::tracks) }
    suspend fun favoritesOrNull(): List<Track>? =
        io { raw(runCatching { Odmobile.favorites() }.getOrNull())?.let(Json::tracks) }
    suspend fun playlistsOrNull(): List<Playlist>? =
        io { raw(runCatching { Odmobile.playlists() }.getOrNull())?.let(Json::playlists) }

    private fun raw(s: String?): String? = if (Json.hasError(s)) null else s

    // ---- radio mixes ----
    // "Song radio" / "artist radio": both return {tracks:[...]} in the same wire
    // shape as Flow, so they parse and play through the exact Flow path.

    suspend fun trackMix(id: String): List<Track> = io { Json.tracks(Odmobile.trackMixJSON(id)) }
    suspend fun artistMix(id: String): List<Track> = io { Json.tracks(Odmobile.artistMixJSON(id)) }

    // ---- listening history (machine-local) ----
    // HistoryRecentJSON returns a bare JSON array (newest first); HistoryStatsJSON
    // returns {topTracks,topArtists,totalSeconds}. gomobile maps Go int -> Java
    // long, hence the toLong() on the argument.

    suspend fun historyRecent(n: Int): List<HistoryEntry> =
        io { Json.historyRecent(runCatching { Odmobile.historyRecentJSON(n.toLong()) }.getOrNull()) }
    suspend fun historyStats(sinceDays: Int): HistoryStats =
        io { Json.historyStats(runCatching { Odmobile.historyStatsJSON(sinceDays.toLong()) }.getOrNull()) }

    // ---- playback ----

    suspend fun play(trackId: String, durationMs: Long): Boolean =
        io { runCatching { Odmobile.play(trackId, durationMs) }.getOrDefault(false) }

    suspend fun playEpisode(id: String, durationMs: Long = 0L): Boolean =
        io { runCatching { Odmobile.playEpisodeMS(id, durationMs) }.getOrDefault(false) }

    // ---- gapless preload ----
    // Resolves and buffers the NEXT queue track so the engine swaps into it
    // seamlessly at the track boundary instead of the UI doing a full network
    // re-resolve (mirrors the desktop GUIs' DZPreload). The bound Go func
    // returns an error, which gomobile surfaces as a thrown exception — mapped
    // to false here so callers can fall back to the plain-resolve advance.
    suspend fun preload(id: String): Boolean =
        io { runCatching { Odmobile.preload(id); true }.getOrDefault(false) }

    // Discards a previously armed preload. Call whenever the upcoming track
    // stops being deterministic (shuffle/repeat toggles, queue edits, stop).
    suspend fun clearPreload() = io { runCatching { Odmobile.clearPreload() }.let {} }

    // ---- engine queue sync (app queue -> engine) ----
    // Mirror the app-owned queue into the engine so remote controllers see the
    // real queue on /status and /next|/prev walk it, and so the engine can own
    // gapless-promote over the synced queue. [json] is a JSON array in the shared
    // list wire shape ({id,name,durationMs,artistLine,artistId,artists,albumName,
    // artworkUrl,explicit}; see PlayerController.queueJson). setQueueJson resets
    // the engine cursor to 0, so always follow it with [setQueueIndex].
    suspend fun setQueueJson(json: String): Boolean =
        io { runCatching { Odmobile.setQueueJSON(json); true }.getOrDefault(false) }
    suspend fun setQueueIndex(i: Int) = io { runCatching { Odmobile.setQueueIndex(i.toLong()) }.let {} }

    // ---- engine queue adoption (engine -> app) ----
    // QueueVersion bumps on every engine-queue CONTENT change (a remote
    // controller editing this device's queue, /play/album, …); cursor moves
    // don't bump it. The poll watches this cheaply (a single long read) and
    // pulls the full [queueSnapshot] only when it moves, so the app can adopt
    // remote queue edits (fixes wrong now-playing + remote-loaded albums
    // stopping after one track). Both are synchronous engine reads (no network).
    fun queueVersion(): Long = runCatching { Odmobile.queueVersion() }.getOrDefault(0L)
    fun queueSnapshot(): QueueSnapshot? =
        Json.queueSnapshot(runCatching { Odmobile.queueJSON() }.getOrNull())

    // ---- engine queue edits (app -> engine; P4 native queue editor) ----
    // Mutate the engine-owned queue directly so remote controllers stay
    // consistent; each bumps QueueVersion, which the PlayerController poll adopts
    // to re-sync the app queue + UI. gomobile maps Go int -> Java long, hence the
    // toLong() on the index arguments.

    // Removes the track at [index] (the engine refuses to drop the playing row).
    suspend fun queueRemove(index: Int) = io { runCatching { Odmobile.queueRemove(index.toLong()) }.let {} }
    // Reorders the queue (from -> to); the cursor/history follow. Out-of-range /
    // from==to are engine-side no-ops.
    suspend fun queueMove(from: Int, to: Int) =
        io { runCatching { Odmobile.queueMove(from.toLong(), to.toLong()) }.let {} }
    // Inserts the wire-shaped [trackJson] (single object or 1-element array)
    // immediately after the current row.
    suspend fun queueInsertNext(trackJson: String) =
        io { runCatching { Odmobile.queueInsertNext(trackJson) }.let {} }

    // ---- offline download (premium-only; requires the stream cache enabled) ----
    // Populates the on-disk media cache with the raw ciphertext for [id] so it
    // can later play with zero network. Returns the engine's raw JSON:
    // {"status":"cached|already_cached|cache_only","key":"...","trackID":"..."}
    // or {"error":"..."} (e.g. "media cache disabled" when the stream cache is
    // off). Blocking / network-bound — pinned to Dispatchers.IO.
    suspend fun downloadForOffline(id: String): String =
        io { runCatching { Odmobile.downloadForOffline(id) }.getOrDefault("""{"error":"failed"}""") }

    // ---- downloads (premium-only; the engine rejects the request otherwise) ----
    // Returns the engine's raw JSON: {"path":"..."} on success or {"error":"..."}.
    // Pass "" for [dir] to save into the shared default download folder.
    suspend fun download(id: String, dir: String): String =
        io { runCatching { Odmobile.downloadTrack(id, dir) }.getOrDefault("""{"error":"failed"}""") }

    // Album/playlist batch download to the shared folder; returns the engine's
    // batch summary {"saved":N,"failed":N,"dir":"...","error":""}. Premium-only,
    // blocking — kept on Dispatchers.IO like [download].
    suspend fun downloadAlbum(id: String): String =
        io { runCatching { Odmobile.downloadAlbum(id) }.getOrDefault("""{"error":"failed"}""") }
    suspend fun downloadPlaylist(id: String): String =
        io { runCatching { Odmobile.downloadPlaylist(id) }.getOrDefault("""{"error":"failed"}""") }

    // The shared default download folder (empty when the engine has none yet).
    suspend fun downloadDir(): String = io { runCatching { Odmobile.downloadDir() }.getOrDefault("") }

    // Persists a new shared default download folder; false when the engine can't use it.
    suspend fun setDownloadDir(p: String): Boolean =
        io { runCatching { Odmobile.setDownloadDir(p) }.getOrDefault(false) }

    // Whether the currently-playing stream is a 30s preview rather than the full track.
    fun isPreview(): Boolean = runCatching { Odmobile.isPreview() }.getOrDefault(false)

    // ---- ads (Deezer Free only) ----
    // Whether ad reporting is suppressed. Disabling ads also stops reporting plays,
    // which breaks Deezer's terms — see the disclaimer in SettingsScreen.
    suspend fun adsDisabled(): Boolean = io { runCatching { Odmobile.adsDisabled() }.getOrDefault(false) }
    suspend fun setAdsDisabled(off: Boolean): Boolean =
        io { runCatching { Odmobile.setAdsDisabled(off) }.getOrDefault(false) }

    fun pause() = runCatching { Odmobile.pause() }.let {}
    fun resume() = runCatching { Odmobile.resume() }.let {}
    fun togglePause() = runCatching { Odmobile.togglePause() }.let {}
    fun stop() = runCatching { Odmobile.stop() }.let {}
    fun seek(ms: Long) = runCatching { Odmobile.seek(ms) }.let {}
    fun setVolume(v: Double) = runCatching { Odmobile.setVolume(v) }.let {}

    // Suspends/resumes the local OS audio device without touching playback
    // state — used for audio-focus handling (e.g. release audio during calls).
    fun setOutputSuspended(on: Boolean) = runCatching { Odmobile.setOutputSuspended(on) }.let {}

    fun volume(): Double = runCatching { Odmobile.volume() }.getOrDefault(1.0)
    fun state(): Int = runCatching { Odmobile.state().toInt() }.getOrDefault(STOPPED)
    fun positionMs(): Long = runCatching { Odmobile.positionMS() }.getOrDefault(0L)
    fun durationMs(): Long = runCatching { Odmobile.durationMS() }.getOrDefault(0L)
    fun format(): String = runCatching { Odmobile.format() }.getOrDefault("")
    fun finishedCount(): Int = runCatching { Odmobile.finishedCount().toInt() }.getOrDefault(0)
    fun nowPlaying(): Track? = Json.nowPlaying(runCatching { Odmobile.nowPlaying() }.getOrNull())

    // ---- settings ----

    fun setQuality(level: Int) = runCatching { Odmobile.setQuality(level.toLong()) }.let {}
    fun quality(): Int = runCatching { Odmobile.quality().toInt() }.getOrDefault(0)
    fun setReplayGain(on: Boolean) = runCatching { Odmobile.setReplayGain(on) }.let {}
    fun replayGain(): Boolean = runCatching { Odmobile.replayGain() }.getOrDefault(false)
    fun setGapless(on: Boolean) = runCatching { Odmobile.setGapless(on) }.let {}
    fun gapless(): Boolean = runCatching { Odmobile.gapless() }.getOrDefault(true)
    fun setCrossfadeMs(ms: Int) = runCatching { Odmobile.setCrossfadeMS(ms.toLong()) }.let {}
    fun crossfadeMs(): Int = runCatching { Odmobile.crossfadeMS().toInt() }.getOrDefault(0)

    // Raw-stream on-disk cache budget in MB (0 = off, the default). The cache is
    // attached to the player once at startup, so a change only takes effect at the
    // next launch — persist it in Prefs and re-apply on login like gapless/crossfade.
    fun mediaCacheMB(): Int = runCatching { Odmobile.mediaCacheMB().toInt() }.getOrDefault(0)
    fun setMediaCacheMB(mb: Int): Boolean =
        runCatching { Odmobile.setMediaCacheMB(mb.toLong()) }.getOrDefault(false)

    // ---- equalizer + mono downmix ----
    // The engine owns all EQ state and persistence (manual band edits flip the
    // preset to "custom" engine-side); these only read state / forward changes.

    fun eqState(): EqState? = Json.eqState(runCatching { Odmobile.eqjson() }.getOrNull())
    fun setEqEnabled(on: Boolean) = setEq("""{"enabled":$on}""")
    fun setEqMono(on: Boolean) = setEq("""{"mono":$on}""")
    fun setEqPreset(name: String) = setEq("""{"preset":${JSONObject.quote(name)}}""")
    fun setEqBand(index: Int, gainDb: Double) = setEq("""{"band":{"index":$index,"gainDb":$gainDb}}""")
    fun setEqPreamp(db: Double) = setEq("""{"preampDb":$db}""")

    private fun setEq(js: String) = runCatching { Odmobile.setEQJSON(js) }.let {}

    // ---- sleep timer ----
    // Pause after [minutes] (auto fade-out), or when the current track ends when
    // [endOfTrack] is true. minutes <= 0 with endOfTrack == false cancels the timer.
    fun setSleepTimer(minutes: Int, endOfTrack: Boolean) =
        runCatching { Odmobile.setSleepTimer(minutes.toLong(), if (endOfTrack) 1L else 0L) }.let {}
    fun cancelSleepTimer() = runCatching { Odmobile.cancelSleepTimer() }.let {}
    fun sleepActive(): Boolean = runCatching { Odmobile.sleepActive() != 0L }.getOrDefault(false)
    fun sleepEndOfTrack(): Boolean = runCatching { Odmobile.sleepEndOfTrack() != 0L }.getOrDefault(false)
    fun sleepRemainingMs(): Long = runCatching { Odmobile.sleepRemainingMS() }.getOrDefault(0L)

    // ---- library writes ----

    suspend fun addFavorite(id: String): Boolean = io { runCatching { Odmobile.addFavorite(id) }.getOrDefault(false) }
    suspend fun removeFavorite(id: String): Boolean = io { runCatching { Odmobile.removeFavorite(id) }.getOrDefault(false) }
    suspend fun addToPlaylist(playlistId: String, trackId: String): Boolean =
        io { runCatching { Odmobile.addToPlaylist(playlistId, trackId) }.getOrDefault(false) }
    suspend fun removeFromPlaylist(playlistId: String, trackId: String): Boolean =
        io { runCatching { Odmobile.removeFromPlaylist(playlistId, trackId) }.getOrDefault(false) }
    suspend fun createPlaylist(title: String): String? = io { Json.playlistId(Odmobile.createPlaylist(title)) }
    suspend fun renamePlaylist(id: String, title: String): Boolean =
        io { runCatching { Odmobile.renamePlaylist(id, title) }.getOrDefault(false) }
    suspend fun deletePlaylist(id: String): Boolean = io { runCatching { Odmobile.deletePlaylist(id) }.getOrDefault(false) }

    // ---- OpenDeezer Connect ----

    suspend fun discoverDevices(timeoutMs: Long = 700L): List<ConnectDevice> =
        io { Json.devices(Odmobile.discoverDevices(timeoutMs)) }
    suspend fun connectDevice(addr: String): Boolean =
        io { runCatching { Odmobile.connectDevice(addr) }.getOrDefault(false) }
    // Suspend: sends a final Stop to the remote over HTTP, so keep it off main.
    suspend fun disconnectDevice() = io { runCatching { Odmobile.disconnectDevice() }.let {} }
    fun connectedDevice(): String = runCatching { Odmobile.connectedDevice() }.getOrDefault("")

    // ---- repeat / shuffle (forwarded to remote when a Connect device is active) ----

    // mode: 0=off, 1=all, 2=one
    fun setRepeat(mode: Int) = runCatching { Odmobile.setRepeat(mode.toLong()) }.let {}
    // on: 0=off, 1=on
    fun setShuffle(on: Int) = runCatching { Odmobile.setShuffle(on.toLong()) }.let {}

    // Engine-owned repeat/shuffle truth. While routed to a Connect device these
    // return the remote host's mode (its routed snapshot), so the poll reflects
    // the host's real mode when casting and picks up external/remote changes.
    // mode: 0=off, 1=all, 2=one.
    fun repeatMode(): Int = when (runCatching { Odmobile.getRepeat() }.getOrDefault("off")) {
        "all" -> 1
        "one" -> 2
        else -> 0
    }
    fun shuffleOn(): Boolean = runCatching { Odmobile.getShuffle() }.getOrDefault(false)

    // ---- phone web remote ----

    fun setWebRemoteEnabled(on: Boolean) =
        runCatching { Odmobile.webRemoteSetEnabled(if (on) 1L else 0L) }.let {}

    fun webRemoteInfo(): WebRemoteInfo? =
        Json.webRemoteInfo(runCatching { Odmobile.webRemoteInfo() }.getOrNull())

    suspend fun webRemoteQRPng(): ByteArray =
        io { runCatching { Odmobile.webRemoteQRPNG() ?: ByteArray(0) }.getOrDefault(ByteArray(0)) }

    // ---- OpenDeezer Connect host ----
    // Advertise this device so other same-account OpenDeezer apps can discover
    // and control it. Mirrors the desktop "make this device reachable" toggle.

    fun setConnectHostEnabled(on: Boolean) =
        runCatching { Odmobile.connectHostSetEnabled(if (on) 1L else 0L) }.let {}

    fun connectHostInfo(): ConnectHostInfo? =
        Json.connectHostInfo(runCatching { Odmobile.connectHostInfo() }.getOrNull())

    private suspend inline fun <T> io(crossinline block: () -> T): T =
        withContext(Dispatchers.IO) { block() }
}

File: gui/ios/Sources/OpenDeezer/Engine.swift
import Foundation
import Odmobile

/// Errors surfaced by the engine layer to the UI.
enum EngineError: LocalizedError {
    case server(String)
    case decode(String)

    var errorDescription: String? {
        switch self {
        case .server(let message): return message
        case .decode(let message): return String(localized: "Couldn't read the server's response (\(message)).")
        }
    }
}

/// Thin async/await façade over the `Odmobile*` C functions gomobile generates
/// from `mobile/odmobile.go`. Every call that can block (network I/O in the Go
/// runtime) is dispatched onto a background serial queue so SwiftUI's main
/// thread never stalls; state getters that are simple field reads are called
/// directly.
enum Engine {
    private static let ioQueue = DispatchQueue(label: "fr.cyclooo.OpenDeezer.engine.io", qos: .userInitiated)
    /// Bulk artwork downloads run here, concurrently, so a screenful of slow
    /// cover fetches (15s HTTP timeout each) can't head-of-line-block
    /// play/pause/seek on the serial transport queue.
    private static let fetchQueue = DispatchQueue(label: "fr.cyclooo.OpenDeezer.engine.fetch", qos: .utility, attributes: .concurrent)
    private static let decoder = JSONDecoder()

    private static func run<T: Sendable>(_ body: @escaping @Sendable () -> T) async -> T {
        await withCheckedContinuation { continuation in
            ioQueue.async { continuation.resume(returning: body()) }
        }
    }

    private static func runFetch<T: Sendable>(_ body: @escaping @Sendable () -> T) async -> T {
        await withCheckedContinuation { continuation in
            fetchQueue.async { continuation.resume(returning: body()) }
        }
    }

    private static func decode<T: Decodable>(_ json: String, as type: T.Type) throws -> T {
        let data = Data(json.utf8)
        if let err = try? decoder.decode(ErrorResponse.self, from: data), !err.error.isEmpty {
            throw EngineError.server(err.error)
        }
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw EngineError.decode(error.localizedDescription)
        }
    }

    // MARK: - Lifecycle / account

    static func initEngine(arl: String) async -> Bool { await run { OdmobileInit(arl) } }
    static func loggedIn() async -> Bool { await run { OdmobileLoggedIn() } }
    /// Why the most recent `initEngine` returned false: 0 ok, 1 ARL expired or
    /// invalid, 2 no internet, 3 other. A cheap atomic read engine-side, so
    /// (like `state`/`volume`) it's called directly; only meaningful right after
    /// `initEngine` returns false.
    static func loginErrorKind() -> Int { OdmobileLoginErrorKind() }
    static func setClientInfo(client: String, device: String) { OdmobileSetClientInfo(client, device) }

    static func account() async throws -> Account {
        try decode(await run { OdmobileAccount() }, as: Account.self)
    }

    /// Checks GitHub for a newer release. Network failures decode to
    /// `hasUpdate == false` engine-side, so this rarely throws in practice.
    static func checkUpdate() async -> UpdateInfo? {
        try? decode(await run { OdmobileCheckUpdate() }, as: UpdateInfo.self)
    }

    /// Tears the engine session down: stops playback, closes the control
    /// server (web-remote pairing + same-account auth) and the Connect-host
    /// advertiser, and forgets the Deezer client. A later `initEngine` starts
    /// services fresh for the new account.
    static func logout() async { await run { OdmobileLogout() } }

    // MARK: - Browse

    static func home() async throws -> HomeResponse {
        try decode(await run { OdmobileHome() }, as: HomeResponse.self)
    }
    static func favorites() async throws -> [Track] {
        try decode(await run { OdmobileFavorites() }, as: TracksResponse.self).tracks
    }
    static func playlists() async throws -> [Playlist] {
        try decode(await run { OdmobilePlaylists() }, as: PlaylistsResponse.self).playlists
    }
    static func playlistTracks(_ id: String) async throws -> [Track] {
        try decode(await run { OdmobilePlaylistTracks(id) }, as: TracksResponse.self).tracks
    }
    static func albumTracks(_ id: String) async throws -> [Track] {
        try decode(await run { OdmobileAlbumTracks(id) }, as: TracksResponse.self).tracks
    }
    static func flow() async throws -> [Track] {
        try decode(await run { OdmobileFlow() }, as: TracksResponse.self).tracks
    }
    static func charts() async throws -> ChartsResponse {
        try decode(await run { OdmobileCharts() }, as: ChartsResponse.self)
    }
    static func search(_ query: String) async throws -> SearchResponse {
        try decode(await run { OdmobileSearch(query) }, as: SearchResponse.self)
    }
    static func artistTop(_ id: String) async throws -> [Track] {
        try decode(await run { OdmobileArtistTop(id) }, as: TracksResponse.self).tracks
    }
    static func artistProfile(_ id: String) async throws -> ArtistProfilePage {
        try decode(await run { OdmobileArtistProfile(id) }, as: ArtistProfilePage.self)
    }
    static func lyrics(_ id: String) async throws -> Lyrics {
        try decode(await run { OdmobileLyrics(id) }, as: Lyrics.self)
    }
    static func searchPodcasts(_ query: String) async throws -> [Podcast] {
        try decode(await run { OdmobileSearchPodcasts(query) }, as: PodcastsResponse.self).podcasts
    }
    static func podcastEpisodes(_ id: String) async throws -> [Episode] {
        try decode(await run { OdmobilePodcastEpisodes(id) }, as: EpisodesResponse.self).episodes
    }

    // MARK: - Playback

    static func play(id: String, durationMs: Int64) async -> Bool {
        await run { OdmobilePlay(id, durationMs) }
    }
    static func playEpisode(id: String, durationMs: Int64) async -> Bool {
        await run { OdmobilePlayEpisodeMS(id, durationMs) }
    }
    /// Resolves and buffers the next queue track so the engine swaps into it
    /// seamlessly at the track boundary (mirrors the desktop GUIs' DZPreload).
    /// False on failure — the caller then falls back to a plain re-resolve at
    /// the boundary. `OdmobilePreload` is the gomobile binding of
    /// `Preload(id string) error` (BOOL + NSError out-param).
    static func preload(id: String) async -> Bool {
        await run {
            var err: NSError?
            return OdmobilePreload(id, &err) && err == nil
        }
    }
    /// Discards a previously armed preload. Call whenever the upcoming track
    /// stops being deterministic (shuffle/repeat toggles, queue edits, stop).
    static func clearPreload() async { await run { OdmobileClearPreload() } }
    static func pause() async { await run { OdmobilePause() } }
    static func resume() async { await run { OdmobileResume() } }
    static func togglePause() async { await run { OdmobileTogglePause() } }
    static func stop() async { await run { OdmobileStop() } }
    static func seek(ms: Int64) async { await run { OdmobileSeek(ms) } }
    /// Async: when routed to a Connect device this becomes a blocking HTTP
    /// POST (15s timeout) engine-side — it must never run on the main thread.
    static func setVolume(_ v: Double) async { await run { OdmobileSetVolume(v) } }
    /// Suspends/resumes the local OS audio device (never Connect-routed);
    /// cheap and local, safe to call synchronously.
    static func setOutputSuspended(_ on: Bool) { OdmobileSetOutputSuspended(on) }
    static func volume() -> Double { OdmobileVolume() }
    static func state() -> Int { OdmobileState() }
    static func positionMS() -> Int64 { OdmobilePositionMS() }
    static func durationMS() -> Int64 { OdmobileDurationMS() }
    static func finishedCount() -> Int { OdmobileFinishedCount() }
    static func format() -> String { OdmobileFormat() }

    static func nowPlaying() async -> Track? {
        let json = await run { OdmobileNowPlaying() }
        guard let track = try? decode(json, as: Track.self), !track.id.isEmpty else { return nil }
        return track
    }

    // MARK: - Library writes

    static func addFavorite(_ id: String) async -> Bool { await run { OdmobileAddFavorite(id) } }
    static func removeFavorite(_ id: String) async -> Bool { await run { OdmobileRemoveFavorite(id) } }
    static func addToPlaylist(_ playlistID: String, trackID: String) async -> Bool {
        await run { OdmobileAddToPlaylist(playlistID, trackID) }
    }
    static func removeFromPlaylist(_ playlistID: String, trackID: String) async -> Bool {
        await run { OdmobileRemoveFromPlaylist(playlistID, trackID) }
    }
    static func createPlaylist(_ title: String) async throws -> String {
        try decode(await run { OdmobileCreatePlaylist(title) }, as: CreatedPlaylist.self).id
    }
    static func renamePlaylist(_ id: String, title: String) async -> Bool {
        await run { OdmobileRenamePlaylist(id, title) }
    }
    static func deletePlaylist(_ id: String) async -> Bool { await run { OdmobileDeletePlaylist(id) } }

    // MARK: - Settings

    static func setQuality(_ level: Int) { OdmobileSetQuality(level) }
    static func quality() -> Int { OdmobileQuality() }
    static func setReplayGain(_ on: Bool) { OdmobileSetReplayGain(on) }
    static func replayGain() -> Bool { OdmobileReplayGain() }
    static func setGapless(_ on: Bool) { OdmobileSetGapless(on) }
    static func gapless() -> Bool { OdmobileGapless() }
    static func setCrossfadeMS(_ ms: Int) { OdmobileSetCrossfadeMS(ms) }
    static func crossfadeMS() -> Int { OdmobileCrossfadeMS() }

    // MARK: - Equalizer

    /// Full EQ + mono-downmix state. The DSP, persistence and the
    /// preset→"custom" flip on manual band edits all live engine-side; the UI
    /// only renders this state and forwards changes.
    static func eqState() async -> EQState? {
        let json = await run { OdmobileEQJSON() }
        return try? decode(json, as: EQState.self)
    }

    /// EQ setters are cheap in-memory DSP-parameter writes (the engine
    /// debounces its own persistence), so they're safe to call synchronously.
    static func setEQEnabled(_ on: Bool) { applyEQ(["enabled": on]) }
    static func setEQMono(_ on: Bool) { applyEQ(["mono": on]) }
    static func setEQPreset(_ name: String) { applyEQ(["preset": name]) }
    static func setEQBand(index: Int, gainDb: Double) { applyEQ(["band": ["index": index, "gainDb": gainDb]]) }
    static func setEQPreamp(_ db: Double) { applyEQ(["preampDb": db]) }

    /// Serializes a partial EQ update for `OdmobileSetEQJSON` (every key optional).
    private static func applyEQ(_ fields: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: fields),
              let json = String(data: data, encoding: .utf8) else { return }
        _ = OdmobileSetEQJSON(json)
    }

    // MARK: - Sleep timer

    /// Pause after `minutes` (with fade-out), or when the current track ends if
    /// `endOfTrack` is true. Passing `minutes <= 0` with `endOfTrack == false`
    /// cancels any pending timer.
    static func setSleepTimer(minutes: Int, endOfTrack: Bool) {
        OdmobileSetSleepTimer(minutes, endOfTrack ? 1 : 0)
    }
    static func cancelSleepTimer() { OdmobileCancelSleepTimer() }
    static func sleepActive() -> Bool { OdmobileSleepActive() != 0 }
    static func sleepEndOfTrack() -> Bool { OdmobileSleepEndOfTrack() != 0 }
    static func sleepRemainingMS() -> Int64 { OdmobileSleepRemainingMS() }

    // MARK: - Connect

    static func discoverDevices(timeoutMs: Int) async -> [Device] {
        let json = await run { OdmobileDiscoverDevices(timeoutMs) }
        return (try? decode(json, as: [Device].self)) ?? []
    }
    static func connectDevice(_ addr: String) async -> Bool { await run { OdmobileConnectDevice(addr) } }
    static func disconnectDevice() async { await run { OdmobileDisconnectDevice() } }
    static func connectedDevice() -> String { OdmobileConnectedDevice() }
    /// Async like setVolume: Connect-routed, so a blocking HTTP POST engine-side.
    static func setRepeat(_ mode: Int) async { await run { OdmobileSetRepeat(mode) } }
    static func setShuffle(_ on: Bool) async { await run { OdmobileSetShuffle(on ? 1 : 0) } }

    // MARK: - Web remote

    static func webRemoteSetEnabled(_ on: Bool) { OdmobileWebRemoteSetEnabled(on ? 1 : 0) }
    static func webRemoteInfo() async -> WebRemoteInfo? {
        let json = await run { OdmobileWebRemoteInfo() }
        return try? decode(json, as: WebRemoteInfo.self)
    }
    static func webRemoteQRPNG() async -> Data? { await run { OdmobileWebRemoteQRPNG() } }

    // MARK: - OpenDeezer Connect host (make this device controllable)

    static func connectHostSetEnabled(_ on: Bool) { OdmobileConnectHostSetEnabled(on ? 1 : 0) }
    static func connectHostInfo() async -> ConnectHostInfo? {
        let json = await run { OdmobileConnectHostInfo() }
        return try? decode(json, as: ConnectHostInfo.self)
    }

    // MARK: - Downloads (premium-only)

    /// `OdmobileDownloadTrack` result: `{"path":"…"}` on success, `{"error":"…"}`
    /// on failure. Both optional so an empty/malformed reply still decodes.
    struct DownloadResult: Decodable {
        let path: String?
        let error: String?
    }

    /// Downloads `id` into `destDir` ("" -> the engine's shared default folder).
    /// Blocking engine-side (network fetch + Blowfish decrypt), so it runs on the
    /// IO queue. Downloads are premium-only; a Free account comes back with an
    /// error in the result.
    static func download(id: String, destDir: String = "") async -> DownloadResult {
        let json = await run { OdmobileDownloadTrack(id, destDir) }
        return (try? decoder.decode(DownloadResult.self, from: Data(json.utf8)))
            ?? DownloadResult(path: nil, error: nil)
    }
    /// The current download folder (env / config / default).
    static func downloadDir() async -> String { await run { OdmobileDownloadDir() } }
    /// Persists the download folder ("" resets to the default); true on success.
    @discardableResult
    static func setDownloadDir(_ path: String) async -> Bool { await run { OdmobileSetDownloadDir(path) } }
    /// True when the current track is Deezer's 30-second preview fallback — rare;
    /// a Free account normally streams the full 128 kbps track. Cheap engine read,
    /// so (like `state`/`volume`) it's called directly.
    static func isPreview() -> Bool { OdmobileIsPreview() }

    // MARK: - Free-tier ads / play reporting (Deezer Free only)

    /// Whether the Free-tier ads / play-reporting opt-out is on. Only meaningful
    /// for a Deezer Free account — paid plans carry no ads.
    static func adsDisabled() async -> Bool { await run { OdmobileAdsDisabled() } }
    @discardableResult
    static func setAdsDisabled(_ disabled: Bool) async -> Bool { await run { OdmobileSetAdsDisabled(disabled) } }

    // MARK: - Misc

    static func fetch(_ url: String) async -> Data? { await runFetch { OdmobileFetch(url) } }
}
